dali/pipeline/util/batch_utils.h

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_UTIL_BATCH_UTILS_H_
+#define DALI_PIPELINE_UTIL_BATCH_UTILS_H_
+
+#include <memory>
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/pipeline/data/tensor_vector.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+/**
+ * @brief Verifies that the inputs in the workspace satisfy the layout
+ *        constraints imposed by the schema.
+ */
+template <typename Workspace>
+inline void CheckInputLayouts(const Workspace &ws, const OpSpec &spec) {
+  auto &schema = spec.GetSchema();
+  for (int i = 0; i < spec.NumRegularInput(); ++i) {
+    if (ws.template InputIsType<CPUBackend>(i)) {
+      auto &input = ws.template Input<CPUBackend>(i);
+      (void)schema.GetInputLayout(i, input.shape().sample_dim(), input.GetLayout());
+    } else if (ws.template InputIsType<GPUBackend>(i)) {
+      auto &input = ws.template Input<GPUBackend>(i);
+      (void)schema.GetInputLayout(i, input.shape().sample_dim(), input.GetLayout());
+    } else {
+      DALI_FAIL(make_string("Input ", i, " has an unknown backend"));
+    }
+  }
+}
+
+template <typename InputRef>
+bool SetDefaultLayoutIfNeeded(InputRef &in, const OpSchema &schema, int in_idx) {
+  if (!in.GetLayout().empty()) {
+    return false;
+  }
+  auto default_layout = schema.GetInputLayout(in_idx, in.shape().sample_dim(), in.GetLayout());
+  if (default_layout.empty()) {
+    return false;
+  }
+  in.SetLayout(default_layout);
+  return true;
+}
+
+template <typename Backend>
+std::shared_ptr<TensorList<Backend>> PresentAsTensorList(
+    const std::shared_ptr<TensorList<Backend>> &in) {
+  return in;
+}
+
+template <typename Backend>
+std::shared_ptr<TensorList<Backend>> PresentAsTensorList(
+    const std::shared_ptr<TensorVector<Backend>> &in) {
+  if (in->IsContiguous()) {
+    return in->AsTensorList(false);
```

Comment:
I'm not sure if this going to work. `in->AsTensorList(false)` skips the continuity check.

Reply:
I don't think I understand, I skip the continuity check in `AsTensorList` because I just checked it in the if condition.

Reply:
My point is, that in the original usage in workspace_policy.h this check was deliberately skipped, as the code was used to connect inputs and outputs in the graph before the actual TL were allocated, and they were not continuous at that time even if they were allocated as such later.

Reply:
Yes.

Reply:
To elaborate, the problem with this function is that it breaks the encapsulation of the TensorVector and uses that implementation detail to build the graph of inputs-outputs for the execution. We can probably use this as optimization for now in the eager ops (maybe add the comment mentioning it, that we know what we are doing - the TV is already finalized and filled with data, so we can use a shortcut), but the two usages of the function have distinct preconditions - @ksztenderski's usage requires valid and filled TensorVector, the current usage of PresentAsTensorList is a hack for executor to give the next operator the access to contiguous memory shown as TensorList that will be filled with data, before TensorVector is actually used. 

They should not have been generalized together.

Reply:
Reverted to separate implementations of `PresentAsTensorList` and `AsTensorList` for eager operator.